README.md
# durable-persistence

This repository is for exploration of persistent datastructures on a durable
medium. The durable medium is abstracted away by a minimalistic
[key-value protocol](https://github.com/replikativ/konserve) providing atomic
operations (ACID) on single keys. This decouples the implementation of the
persistent datastructure from the host and backend store.

The repository contains an adaptation of the
[cookbook in-memory version](https://github.com/clojure-cookbook/clojure-cookbook/blob/master/02_composite-data/2-27_and_2-28_custom-data-structures/2-27_red-black-trees-part-i.asciidoc)
of a red black tree. Instead of keeping all fragments in memory we introduce a
Reference type to fragments in the store and dynamically load and store
fragments. The root can be assigned to a fixed key, modelling a durable identity
similar to a Clojure atom.

The long term motivation is to provide efficient indices as a building block for
[datatype management](https://github.com/replikativ/replikativ) independent of
JVM IO libraries spanning also to ClojureScript and the Browser. With konserve
an IndexedDB storage backend for the Browser already exists. For this reason IO
should happen asynchronously through `core.async` in `konserve`. A durable Index
for DataScript would be nice :).

## Usage

The easiest is to check out the project and hack it itself, but if you just want
to use it, you can also add this dependency:
[![Clojars Project](http://clojars.org/io.replikativ/durable-persistence/latest-version.svg)](http://clojars.org/io.replikativ/durable-persistence)



Here is a small benchmark and usage example:

~~~clojure
(require '[durable-persistence.redblack :refer :all]
         '[konserve.core :as k]
         '[konserve.filestore :refer [new-fs-store]]
         '[hasch.core :refer [uuid]]
         '[clojure.core.async :refer [go <!! <!]])


(def store (<!! (new-fs-store "/tmp/async-tree-experiment"
                              :read-handlers (atom {'async-trees.redblack.Ref map->Ref}))))

(def rb-tree (time (reduce (fn [rb-tree elem]
                             (<!! (go
                                    (let [st (System/currentTimeMillis)
                                          rb-tree (<! (insert-val store rb-tree elem))]
                                      (<! (k/assoc-in store [elem] [(uuid) [(uuid)]])) ;; add some data fragment
                                      (<! (k/assoc-in store [:root] rb-tree))
                                      (when (= (mod elem 1000) 0)
                                        (let [delta (- (System/currentTimeMillis)
                                                       st)]
                                          (println "Op for" elem " took " delta " ms")))
                                      rb-tree)))) nil (range 20000))))

;; printouts on the REPL
Op for 0  took  6  ms
Op for 1000  took  48  ms
Op for 2000  took  49  ms
Op for 3000  took  55  ms
Op for 4000  took  48  ms
Op for 5000  took  60  ms
Op for 6000  took  65  ms
Op for 7000  took  61  ms
Op for 8000  took  63  ms
Op for 9000  took  66  ms
Op for 10000  took  62  ms
Op for 11000  took  34  ms
Op for 12000  took  65  ms
Op for 13000  took  63  ms
Op for 14000  took  65  ms
Op for 15000  took  55  ms
Op for 16000  took  60  ms
Op for 17000  took  77  ms
Op for 18000  took  67  ms
Op for 19000  took  69  ms
"Elapsed time: 1239405.719974 msecs"
~~~

We can do efficient range queries on the index, too.

~~~clojure
(time (= (<!! (range-vals store rb-tree -1 20000))
              (range 20000)))
=> true
"Elapsed time: 7565.761136 msecs"
~~~

Shuffeling the inputs is better than worse case (in sequence):
~~~clojure
Op for 6000  took  33  ms
Op for 5000  took  22  ms
Op for 8000  took  35  ms
Op for 16000  took  39  ms
Op for 15000  took  33  ms
Op for 10000  took  31  ms
Op for 4000  took  38  ms
Op for 2000  took  37  ms
Op for 14000  took  47  ms
Op for 19000  took  53  ms
Op for 12000  took  29  ms
Op for 1000  took  45  ms
Op for 18000  took  51  ms
Op for 9000  took  25  ms
Op for 7000  took  48  ms
Op for 17000  took  56  ms
Op for 13000  took  44  ms
Op for 3000  took  50  ms
Op for 11000  took  56  ms
Op for 0  took  42  ms
"Elapsed time: 793967.016349 msecs"
~~~

## Garbage collection

20000 integer inserts cause almost 500 MiB garbage, because we repeatedly write
the index fragments and they have an edn serialization. This persistency
approach similar to Clojure's in memory one provides concurrency for operations
on multiple roots of shared fragments of the tree.

An approach along the lines of a mark and sweep collector together with a
monotonic timestamp on indices should provide short stop the world phases. This
needs to be done for a practical feasibility of course, but first the proper
datastructures should be found.

A copying garbage collector (to another store from the root) is easy to
implement already.

## Metadata format

This is reducible in size still, but focus at the moment is to find efficient
implementations without too much microoptimizations. Exploration of the design
space is still necessary.

## License

Copyright © 2016 Christian Weilbach

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-09-08
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-09-08
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/durable-persistence/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/durable-persistence/compare/0.1.0...0.1.1


test/durable_persistence/core_test.clj
(ns durable-persistence.core-test
  (:require [clojure.test :refer :all]
            [durable-persistence.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/durable_persistence/redblack.clj
(ns durable-persistence.redblack
  (:require [clojure.core.async :refer [chan <! go <!!] :as async]
            [konserve.core :as k]
            [konserve.memory :refer [new-mem-store]]
            [konserve.filestore :refer [new-fs-store]]
            [hasch.core :refer [uuid]]
            [clojure.core.match :refer [match]]
            [clojure.set :as set]
            [clojure.core.cache :as cache]))

;; thankfully adapted from https://github.com/clojure-cookbook/clojure-cookbook/blob/master/02_composite-data/2-27_and_2-28_custom-data-structures/2-27_red-black-trees-part-i.asciidoc


(def C (atom (cache/lru-cache-factory {} :threshold 2048)))

(defrecord Ref [id])

(defn create-ref [store node]
  (go
   (when node
     (if (:id node) ;; TODO ??? (= (type node) async_tree_experiment.core.Ref)
       node
       (let [id (uuid node)]
         (when-not (<! (k/exists? store id))
           (swap! C assoc id node)
           (<! (k/assoc-in store [id] node)))
         (map->Ref {:id id}))))))


(defn load-tree-fragment [store tree depth]
  (go
   (if (= depth 0)
     tree
     (when tree
       (let [[color a y b]
             (if (:id tree)
               (let [v (or (get @C (:id tree))
                           (<! (k/get-in store [(:id tree)])))]
                 (when-not v
                   (throw (ex-info "Value should be in store:" tree)))
                 (swap! C assoc (:id tree) v)
                 v)
               tree)
             new-tree [color
                       (<! (load-tree-fragment store a (dec depth)))
                       y
                       (<! (load-tree-fragment store b (dec depth)))] ]
         new-tree)))))

(defn balance
  "Ensures the given subtree stays balanced by rearranging black nodes
  that have at least one red child and one red grandchild"
  [store tree depth]
  ;; fetch tree to grandchildren
  (go
   (let [tree (<! (load-tree-fragment store tree 2))]
     (match [tree]
            [(:or ;; Left child red with left red grandchild
              [:b [:r [:r a x b] y c] z d]
              ;; Left child red with right red grandchild
              [:b [:r a x [:r b y c]] z d]
              ;; Right child red with left red grandchild
              [:b a x [:r [:r b y c] z d]]
              ;; Right child red with right red grandchild
              [:b a x [:r b y [:r c z d]]])]
            ;; lets write some of the structure to disk here
            (if (= (mod depth 3) 0)
              (let [ar (<! (create-ref store a))
                    br (<! (create-ref store b))
                    cr (<! (create-ref store c))
                    dr (<! (create-ref store d))]
                #_(println "Writing at depth" depth " rebalance")
                [:r [:b ar x br] y [:b cr z dr]])
              [:r [:b a x b] y [:b c z d]])

            [[c1 [c2 a x b] y [c3 c z d]]]
            (if (= (mod depth 3) 0)
              (let [ar (<! (create-ref store a))
                    br (<! (create-ref store b))
                    cr (<! (create-ref store c))
                    dr (<! (create-ref store d))]
                #_(println "Writing at depth" depth)
                [c1 [c2 ar x br] y [c3 cr z dr]])
              [c1 [c2 a x b] y [c3 c z d]])

            :else tree))))

(defn insert-val
  "Inserts x in tree.
  Returns a node with x and no children if tree is nil.

  Returned tree is balanced. See also `balance`"
  [store tree x]
  (go
   (let [ins (fn ins [tree depth]
               (go
                (let [tree (<! (load-tree-fragment store tree 1))]
                  (match tree
                         nil [:r nil x nil]
                         [color a y b] (cond
                                         (< x y)
                                         (<! (balance store [color (<! (ins a (inc depth))) y b]
                                                      depth))
                                         (> x y)
                                         (<! (balance store [color a y (<! (ins b (inc depth)))]
                                                      depth))
                                         :else tree)))))
         [_ a y b] (<! (ins tree 1))]
     [:b a y b])))

(defn find-val
  "Finds value x in tree"
  [store tree x]
  (go
   (let [tree (<! (load-tree-fragment store tree 1))]
     (match tree
            nil nil
            [_ a y b]
            (cond
              (< x y) (<! (find-val store a x))
              (> x y) (<! (find-val store b x))
              :else x)))))

(defn range-vals
  [store tree s e]
  (go
    (let [tree (<! (load-tree-fragment store tree 1))]
      (match tree
             nil []
             [_ a y b]
             (cond
               (and (< s y) (> e y))
               (concat (<! (range-vals store a s e))
                       [y]
                       (<! (range-vals store b s e)))
               (< s y) (<! (range-vals store a s e))
               (> e y) (<! (range-vals store b s e))
               :else [])))))


(comment

  (def store (<!! #_(new-mem-store)
                  (new-fs-store "/tmp/async-tree-experiment"
                                :read-handlers (atom {'async_tree_experiment.core.Ref map->Ref}))))
  (count @assocs)

  (def perf-log (atom []))

  (def inserts (range 20000) #_(shuffle (range 20000)))

  (def rb-tree (time (reduce (fn [rb-tree elem]
                               (<!! (go
                                     (let [st (System/currentTimeMillis)
                                           rb-tree (<! (insert-val store rb-tree elem))]
                                       #_(<! (k/assoc-in store [elem] [(uuid) [(uuid)]]))
                                       (<! (k/assoc-in store [:root] rb-tree))
                                       #_(when-not (<??  (find-val store rb-tree elem))
                                         (throw (ex-info "Oooops." {:failed-to-insert elem
                                                                    :rb-tree rb-tree})))
                                       (when (= (mod elem 1000) 0)
                                         (let [delta (- (System/currentTimeMillis)
                                                        st)]
                                           (swap! perf-log conj delta)
                                           (println "Op for" elem " took " delta " ms")))
                                       rb-tree)))) nil inserts)))

  (time (= (<!! (range-vals store rb-tree -1 20000))
           (range 20000)))

  (count (pr-str (<!! (k/get-in store [:root]))))

  (count (pr-str rb-tree))
  ;; -> [:black [:black nil 0 nil] 1 [:black nil 2 [:red nil 3 nil]]]

  (<!! (find-val store rb-tree 2))
  ;; -> 2

  (time (<!! (find-val store rb-tree "hello")))
  ;; -> nil


  (=
   (uuid [:black nil 64 nil])

   (uuid [:black nil 72 nil])

   (uuid [:black nil 97 nil]))


  (def rb-tree (time (<?? (insert-val store rb-tree 5))))

  (time
   (doseq [i (range 10000)]
     (let [v (<?? (find-val store rb-tree i))]
       (if-not (= i v)
         (throw (ex-info "not equal" {:i i :stored v}))))))



  (defn read-rb-tree-values [rb-tree]
    (go-try
     (let [tree (<? (load-tree rb-tree))
           [_ a y b] tree]
       (match tree
              [_ nil nil nil] #{}
              [_ nil y nil] #{y}
              [_ nil y b] (set/union #{y} (<? (read-rb-tree-values b)))
              [_ a y nil] (set/union #{y} (<? (read-rb-tree-values a)))
              [_ a y b] (set/union (<? (read-rb-tree-values a)) #{y} (<? (read-rb-tree-values b)))))))

  (def values (<?? (read-rb-tree-values rb-tree)))

  (count values)

  (<?? (count-rb-tree rb-tree))




  )


